design/embedded-cluster-cli-driven-update.md

```diff
@@ -0,0 +1,278 @@
+- **Resume capability**: Support interrupted download resumption for reliability
+- **Temporary storage**: Download to temp location before replacement
+
+#### Atomic Replacement Strategy
+Use proven `selfupdate` library approach (same as Replicated CLI):
+- **Backup creation**: Always create backup of current binary before replacement
+- **Atomic operation**: Use filesystem moves/renames for atomicity where possible
+- **Permission preservation**: Maintain original executable permissions and ownership
+- **Cross-platform support**: Handle platform-specific file replacement quirks
+- **Verification post-replacement**: Confirm new binary works correctly
+
+#### Error Handling and Rollback
+Comprehensive error recovery mechanisms:
+- **Download failures**: Network timeouts, authentication errors, disk space issues
+- **Verification failures**: Checksum mismatches, corrupted downloads
+- **Replacement failures**: Permission errors, file locks, disk full
+- **Automatic rollback**: Restore backup if new binary fails basic validation
+- **User guidance**: Clear error messages with suggested remediation steps
+- **Cleanup**: Remove temporary files and failed downloads
+
+#### Security Considerations
+- **HTTPS enforcement**: All downloads over secure connections
+- **License validation**: Verify license before attempting download
+- **File permissions**: Ensure downloaded files have appropriate restrictions
+- **Audit logging**: Log all update attempts for security monitoring
+
+**Implementation Notes:**
+- Mirror Replicated CLI's proven patterns for reliability
+- Handle edge cases like running binary being updated
+- Support both interactive and automated (scripted) usage
+- Graceful degradation when binary update not possible
+
+## Questions
+
+### 1. State Management and API Cursor Support in Stateless Architecture
+
+**Question:** How do we handle state-dependent operations like required release validation, cursor tracking, and deployment history in a stateless CLI environment, and should the versions API endpoint support cursor-based filtering or always return all versions?
+
+**Context:** Embedded cluster's stateless architecture creates challenges for operations that typically require persistent state tracking. Without a local database or persistent storage, we need alternative approaches for:
+- Tracking user's previous update path for required release validation
+- Maintaining cursor positions for incremental API synchronization
+
+### 2. Cross-Channel Semver Support
+
+**Question:** Do we need to account for cross-channel semantic version comparisons, or can we simplify by restricting version comparisons to the same channel?
+
+**Context:** The current design includes cross-channel semver functionality where:
+- Semver-enabled channels can compare versions across different channels (e.g., v1.2.3 from stable > v1.2.1 from beta)
+- Required releases can block updates across ALL channels when semver is enabled
```

Comment:
That's not how it works in KOTS, and I don't think this makes sense either. I can mark the same version / semver as required on a certain channel, but not required on another.